htmd/apps/pmemdlocal.py

```diff
@@ -0,0 +1,183 @@
+from glob import glob as glob
+from shutil import which, move, copy
+from os.path import expanduser
+import threading
+import logging
+import os
+import queue
+import sys
+logger = logging.getLogger(__name__)
+
+
+class PmemdLocal(App):
+    """
+    Parameters
+    ----------
+    pmemd_cuda: str
+        Path to any executable pmemd binary. If None, it will try to detect
+        the pmemd.cuda_SPFP binary
+    ngpus : int
+        Number of GPU devices that PmemdLocal will use. Each simulation will be
+        run on a different GPU. PmemdLocal will use the first `ngpus` devices
+        of the machine.
+    devices : list
+        A list of GPU device indexes on which PmemdLocal is allowed to run
+        simulations. Mutually exclusive with `ngpus`
+    datadir : str
+        A folder to which completed simulations will be moved. If None they
+        will be written in the input directory.
+    """
+    def __init__(self, pmemd_cuda=None, ngpus=None, devices=None, datadir=None):
+        self.states = dict()
+        self.queue = queue.Queue()
+        self.threads = []
+        self.shutdown = False
+        self.system_name = 'Test'
+
+        if ngpus is not None and devices is not None:
+            raise ValueError('Parameters `ngpus` and `devices` are mutually exclusive.')
+
+        if ngpus is None and devices is None:
+            try:
+                devices = range(int(check_output("nvidia-smi -L | wc -l",
+                                shell=True).decode("ascii")))
+            except:
+                raise
+        elif ngpus is not None:
+            devices = range(ngpus)
+
+        if devices is None:
+            raise NameError("""Could not determine which GPUs to use.
+                            Specify the GPUs with the `ngpus=`
+                            or `devices=` parameters""")
+        else:
+            logger.info("Using GPU devices {}".format(','.join(map(str, devices))))
+
+        if not pmemd_cuda:
+            try:
+                # Try to find pmemd.cuda_SPFP in the path
+                pmemd_cuda = which("pmemd.cuda_SPFP", mode=os.X_OK)
+                if pmemd_cuda:
+                    logger.info("Found pmemd.cuda_SPFP at '" + pmemd_cuda + "'")
+            except:
+                raise NameError("""Cannot find 'pmemd.cuda_SPFP' in the PATH
+                                Set its location with the 'pmemd_cuda=' named
+                                argument""")
+
+        if not os.access(pmemd_cuda, os.X_OK):
+            raise NameError("pmemd_cuda file '" + pmemd_cuda + "' not executable")
+
+        self.threads = []
+        for d in devices:
+            t = threading.Thread(target=run_job, args=(self, d, pmemd_cuda,
+                                                       datadir, self.system_name))
+            t.daemon = True
+            t.start()
+            self.threads.append(t)
+
+    # Use the methods from the AcemdLocal class
+    retrieve = AcemdLocal.retrieve
```

Comment:
Hm this suggests to me that these methods should move from `AcemdLocal` to some superclass. Like `AppLocal` which inherits from `App` and implements the queuing system as well.